### PR TITLE
Errors and new API

### PR DESCRIPTION
**Pull request by SirPython**
Created a system for displaying errors in a popup/modal.

Updated the API's `setMessageListener`

### CODE REVIEW

File: src/deck_builder/controller.js [hunk 1]
```diff
@@ -213,6 +216,7 @@ function DeckbuilderController(CardshifterServerAPI, $scope, $rootScope, $locati
 
             $location.path("/game_board");
         } else {
+            ErrorCreator("Not enough cards");
```
Comment (jacwah) [inline src/deck_builder/controller.js]: Shouldn't this be ` ErrorCreator.create("Not enough cards")`?

File: src/lobby/controller.js [hunk 2]
```diff
@@ -122,9 +117,7 @@ function LobbyController(CardshifterServerAPI, $scope, $timeout, $rootScope, $lo
             CardshifterServerAPI.sendMessage(getCards);
             $location.path("/deck_builder");
         } else {
-            var message = new CardshifterServerAPI.messageTypes.ChatMessage(
-                        "Client error: Select a game type before you can open the deck builder.");
-            addChatMessage(message);
+            ErrorCreator.create("Client error: Select a game type before you can open the deck builder.");
```
Comment (jacwah) [inline src/lobby/controller.js]: This message is inconsistent with the other messages. Why does it start with "Client error:"?

File: src/login/controller.js [hunk 3]
```diff
@@ -38,50 +38,56 @@ function LoginController(CardshifterServerAPI, $scope, $location, $rootScope) {
...
                 } catch(e) {
                     // notify the user that there was an issue logging in (loginmessage issue)
+                    ErrorCreator.create("There was an error logging in");
+
                     console.log("LoginMessage error(error 2): " + e);
                     $scope.loggedIn = false;
                     $scope.$apply();
                 }
             } else {
+                ErrorCreator.create("Please enter a username");
```
Comment (jacwah) [inline src/login/controller.js]: This code has too many nesting levels. Maybe things can be separated into functions to remedy this. It's currently really hard to see what `if` this is an `else` to.

File: src/server_interface/server_interface.js [hunk 4]
```diff
@@ -289,26 +290,24 @@ var CardshifterServerAPI = {
...
             var data = JSON.parse(message.data);
-            if(eventTypes) {
-                if(eventTypes.indexOf(data.command) !== -1) { // if contains
-                    listener(data);
-                }
-            } else {
-                listener(data);
+            var command = data.command;
+
+            console.log("Received message", data);
```
Comment (danpantry) [inline src/server_interface/server_interface.js]: remove `console.log` in PRs

File: src/lobby/controller.js [hunk 5]
```diff
@@ -210,6 +203,10 @@ function LobbyController(CardshifterServerAPI, $scope, $timeout, $rootScope, $lo
     function formatTimeNumber(time) {
         return time < 10 ? "0" + time : time;
```
Comment (danpantry) [inline src/lobby/controller.js]: Use `momentjs` instead of reinventing the wheel

Comment (Zomis) [inline src/lobby/controller.js]: Is it really worth adding a new framework/library just to add this feature? I'd prefer to reinvent the very simple wheel in that case.

File: src/server_interface/server_interface.js (hunk 4 above)
Comment (SirPython) [inline src/server_interface/server_interface.js]: The `console.log`s in the API are intentional; they are there for #111.

File: src/lobby/controller.js (hunk 5 above)
Comment (danpantry) [inline src/lobby/controller.js]: Well, momentjs also includes localisation - unless there is no interest in localising Cardshifter?

File: src/server_interface/server_interface.js (hunk 4 above)
Comment (danpantry) [inline src/server_interface/server_interface.js]: Hmm, okay. Then perhaps you could look at using [debug](https://www.npmjs.com/package/debug)?

**Comment by Phrancis** [comment]
Hey @SirPython were you wanting to merge this into `develop` at some point? I just noticed it has been sitting here in conflict since Sept 25

**Pull request merged**